### PR TITLE
LPS-134038 Add missed and remove extra values correctly

### PR DESCRIPTION
**Pull request by mateussgn**
Related issue: https://issues.liferay.com/browse/LPS-134038

### CODE REVIEW

**Comment by mateussgn** [comment]
ci:test:sf

**Comment by mateussgn** [comment]
ci:test:relevant

**Comment by mateussgn** [comment]
The failure is not related to the fix. I'll run the relevant suite again in order to make it ready for forwarding 😃

**Comment by mateussgn** [comment]
ci:test:relevant

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/util/DDMFormValuesConverterUtil.java [hunk 1]
```diff
@@ -40,8 +40,9 @@
 				List<DDMFormFieldValue> nestedDDMFormFieldValues =
 					new ArrayList<>();
 
-				if (StringUtil.equals(ddmFormField.getType(),
-					DDMFormFieldTypeConstants.FIELDSET)) {
+				if (StringUtil.equals(
+						ddmFormField.getType(),
+						DDMFormFieldTypeConstants.FIELDSET)) {
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/util/DDMFormValuesConverterUtil.java:45]: @mateussgn, ideally, this SF should be on the previous commit or on a dedicated one to make it clear to the reviewers and also consistent with the commit message. :wink:

**Review by natocesarrego — CHANGES_REQUESTED** [review]

**Comment by natocesarrego** [comment]
@mateussgn, let's have a call to discuss about this pull tomorrow, I think it'll worth. :slightly_smiling_face:

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/util/DDMFormValuesToFieldsConverterImpl.java [hunk 2]
```diff
@@ -57,16 +54,17 @@ public Fields convert(
 		Map<String, DDMFormField> ddmFormFieldsMap =
 			ddmStructure.getFullHierarchyDDMFormFieldsMap(true);
 
-		List<DDMFormFieldValue> ddmFormFieldValues = _filterDDMFormFieldValues(
-			ddmFormFieldsMap, ddmFormValues.getDDMFormFieldValues());
```
[the remark this answers is not in the document]
Comment (mateussgn) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/util/DDMFormValuesToFieldsConverterImpl.java:61]: This method works perfectly when dealing with field removal from a _structure_ or a _metadata set_, but when dealing with field addition it won't work and makes the web content or documents and media visualization be broken.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/util/DDMFormValuesToFieldsConverterImpl.java [hunk 3]
```diff
@@ -57,16 +54,17 @@ public Fields convert(
...
+		DDMForm ddmForm = ddmStructure.getDDMForm();
+
+		ddmFormValues.setDDMFormFieldValues(
+			DDMFormValuesConverterUtil.addMissingDDMFormFieldValues(
+				ddmForm.getDDMFormFields(),
```
[the remark this answers is not in the document]
Comment (mateussgn) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/util/DDMFormValuesToFieldsConverterImpl.java:61]: They're all used. Nested `ddmFormFields` are inside of `ddmForm.ddmFormFields` and the hierarchy matters.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/util/DDMFormValuesConverterUtil.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	}
+
+	private static void _addNewMissedDDMFormFieldValue(
+		DDMFormField ddmFormField,
+		List<DDMFormFieldValue> ddmFormFieldValueList,
+		List<DDMFormFieldValue> nestedDDMFormFieldValues) {
+
+		DDMFormFieldValue ddmFormFieldValue = new DDMFormFieldValue() {
+			{
+				setInstanceId(StringUtil.randomString());
```
[the remark this answers is not in the document]
Comment (mateussgn) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/util/DDMFormValuesConverterUtil.java:81]: I don't think so. As this code block is responsible for creating e new `ddmFormFieldValue`, the `instanceId` should be different.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/util/DDMFormValuesConverterUtil.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+					}
+				}
+			});
+
+		return ddmFormFieldValueList;
+	}
+
+	private static void _addNewMissedDDMFormFieldValue(
+		DDMFormField ddmFormField,
+		List<DDMFormFieldValue> ddmFormFieldValueList,
```
[the remark this answers is not in the document]
Comment (mateussgn) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/util/DDMFormValuesConverterUtil.java:76]: Sure! I'll do this modification.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/util/DDMFormValuesConverterUtil.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+/**
+ * @author Mateus Santana
+ */
+public class DDMFormValuesConverterUtil {
+
+	public static List<DDMFormFieldValue> addMissingDDMFormFieldValues(
+		List<DDMFormField> ddmFormFields,
+		Map<String, List<DDMFormFieldValue>> ddmFormFieldValues) {
+
+		List<DDMFormFieldValue> ddmFormFieldValueList = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (mateussgn) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/util/DDMFormValuesConverterUtil.java:36]: Perfect! 😄

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/util/DDMFormValuesConverterUtil.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+					_addNewMissedDDMFormFieldValue(
+						ddmFormField, ddmFormFieldValueList,
+						nestedDDMFormFieldValues);
+				}
+				else {
+					if (ListUtil.isNotEmpty(nestedDDMFormFieldValues)) {
+						_addNewMissedDDMFormFieldValue(
+							ddmFormField, ddmFormFieldValueList,
+							nestedDDMFormFieldValues);
+					}
```
[the remark this answers is not in the document]
Comment (mateussgn) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/util/DDMFormValuesConverterUtil.java:63]: Honestly, I didn't get what you mean 😅. However, this call is used in line 53 and reused in line 59.

**Pull request closed without merging**